peers/external_handler.go

```diff
@@ -147,23 +109,99 @@ func (h *RelayerExternalHandler) Disconnected(nodeID ids.NodeID) {
+
+	responseChan := make(chan message.InboundMessage, numExpectedResponses)
+	h.responseChans[requestID] = responseChan
+	h.responsesCount[requestID] = expectedResponses{
+		expected: numExpectedResponses,
+	}
+	return responseChan
+}
+
 // RegisterRequest registers an AppRequest with the timeout manager.
 // If RegisterResponse is not called before the timeout, HandleInbound is called with
 // an internally created AppRequestFailed message.
-func (h *RelayerExternalHandler) RegisterRequest(reqID ids.RequestID) {
+func (h *RelayerExternalHandler) RegisterAppRequest(reqID ids.RequestID) {
 	inMsg := message.InboundAppError(
 		reqID.NodeID,
 		reqID.SourceChainID,
 		reqID.RequestID,
-		common.ErrTimeout.Code,
-		common.ErrTimeout.Message,
+		avalancheCommon.ErrTimeout.Code,
+		avalancheCommon.ErrTimeout.Message,
 	)
 	h.timeoutManager.Put(reqID, false, func() {
 		h.HandleInbound(context.Background(), inMsg)
 	})
 }
 
 // RegisterResponse registers an AppResponse with the timeout manager
-func (h *RelayerExternalHandler) RegisterResponse(reqID ids.RequestID) {
+func (h *RelayerExternalHandler) RegisterAppResponse(inboundMessage message.InboundMessage) {
+	h.responseChansLock.Lock()
+	defer h.responseChansLock.Unlock()
+
+	// Extract the message fields
+	m := inboundMessage.Message()
+
+	// Get the blockchainID from the message.
+	// Note: we should NOT call GetSourceBlockchainID; this is for cross-chain messages using the vm2 interface
+	// For normal app requests messages, the calls result in the same value, but if the relayer handles an
+	// inbound cross-chain app message, then we would get the incorrect chain ID.
+	blockchainID, err := message.GetChainID(m)
+	if err != nil {
+		h.log.Error("Could not get blockchainID from message")
+		inboundMessage.OnFinishedHandling()
+		return
+	}
+	sourceBlockchainID, err := message.GetSourceChainID(m)
+	if err != nil {
+		h.log.Error("Could not get sourceBlockchainID from message")
+		inboundMessage.OnFinishedHandling()
+		return
+	}
+	requestID, ok := message.GetRequestID(m)
+	if !ok {
+		h.log.Error("Could not get requestID from message")
+		inboundMessage.OnFinishedHandling()
+		return
+	}
+
+	reqID := ids.RequestID{
+		NodeID:             inboundMessage.NodeID(),
+		SourceChainID:      sourceBlockchainID,
+		DestinationChainID: blockchainID,
+		RequestID:          requestID,
+		Op:                 byte(inboundMessage.Op()),
+	}
+
+	// Register the response with the timeout manager
 	h.timeoutManager.Remove(reqID)
+
+	// Dispatch to the appropriate response channel
+	if responseChan, ok := h.responseChans[requestID]; ok {
+		responseChan <- inboundMessage
+	} else {
+		h.log.Debug("Could not find response channel for request", zap.Uint32("requestID", requestID))
+	}
+
+	// Check for the expected number of responses, and clear from the map if all expected responses have been received
+	responses := h.responsesCount[reqID.RequestID]
```

Comment:
It's not quite an invariant on `RegisterAppResponse`, since an AppResponse can be sent to the peer for any reason; it doest not necessarily need to be a response to a direct AppRequest. That said, there is a requirement on the responseChan dispatch and responseCount increment that `RegisterRequestID` has been called previously. This can be checked by adding a `return` statement after the debug log when checking if the responseChan for the request ID exists.